### ISSUE TITLE
JSDoc unexpected type error when spreading array containing different shapes of same type

### ISSUE THREAD

**Issue by Arkellys** [microsoft/TypeScript#58437]
### 🔎 Search Terms

"typescript jsdoc error spread union", "jsdoc error spread type different shape"

### 🕗 Version & Regression Information

I was unable to test this on prior versions because the TS playground doesn't handle JSDoc

### ⏯ Playground Link

https://www.typescriptlang.org/play/?ts=5.4.5#code/FASwdgLgpgTgZgQwMZQAQGED2YAmIIjbqoDewqFqSAXKgM4QzgDmA3OZQEYD8tYUAN1jsAvsFCRYiFBmx4C2AEKkOFJL1T8hMdpVSdaWMHBBtgY4BACeABzRGTzVAF5ZufITDKAPm-mf0cSRsBlRoUNcACmDjU0NsRwBtAF0AShcAPlREABs6KHE4AFcwJAUwVBgoCCKYMABBOgBJCAAeABVbKAzIhBhmWk67VMGulQoqmrrUPrMLAHp51ABRAA87MqgceiKkFDo6SygGSMSSfVpzmlQAIiQb1BFHtPZwiFPzg1IL7+u7h6egJe4jep2AAEhPpcqLR-o9HgAaCFQ75fK6w+7wwHAYHARYrdZQTbbRAgHJHE5nZ6pViofEABRgmDsMGsqAA5Jx2agQHRUABbXl0FgU95UtFYl50paM5mwNmc7m8gVCkV4pYAVX4hOJ2QQZNQAHd8AALVAAKQAygARTBIBE7PbHPnGiBm9qW0VgyE-dG3TGApHggB0ocmtQazTFEJ9Er+AcRMZRyZh-oBWIhaRxNKAA

### 💻 Code

The playground link contains the TS version, which work as expected, the below is the JSDoc equivalent throwing the unexpected type error:

```ts
/**
 * @typedef {object} ConditionC
 * @property {string} c
 * @property {never} [b]
 */

/**
 * @typedef {object} ConditionB
 * @property {never} [c]
 * @property {Config} b
 */

/**
 * @typedef {ConditionB | ConditionC} Config
 */

/**
 * @param {Config[]} config
 * @returns {boolean}
 */
const test = (config) => false;

/**
 * @template Type
 * @param {Type} arg
 * @returns {Type}
 */

function returnAsIt(arg) {
  return arg;
}

// Expected success
test([{ b: { c: "c" } }]);
test([{ b: { b: { c: "c" } } }]);

test([
  { b: { c: "c" } },
  { b: { b: { c: "c" } } }
]);

// Expected fail
test([{ }]); // Property 'b' is missing
test([{ b: {} }]); // Property 'b' is missing

// Unexpected fail
test([
  { b: { c: "c" } },
  ...returnAsIt([
    { b: { c: "c" } },
    { b: { b: { c: "c" } } }
  ])
]);
```

### 🙁 Actual behavior

With union type and recursion, when using a function that returns the same type as its input, an unexpected type error appear in some cases.

**These pass:**

```ts
test([
  { b: { c: "c" } },
  ...returnAsIt([
    { b: { c: "c" } }
    { b: { c: "c" } }
  ])
]);
```

```ts
test([
  { b: { c: "c" } },
  ...returnAsIt([
     { b: { b: { c: "c" } } }
     { b: { b: { c: "c" } } }
  ])
]);
```

```ts
test([
  { b: { c: "c" } },
  { b: { b: { c: "c" } } },
  ...returnAsIt([
    { b: { c: "c" } },
    { b: { b: { c: "c" } } }
  ])
]);
```

```ts
test([
  { b: { c: "c" } },
  ...([
    { b: { c: "c" } },
    { b: { b: { c: "c" } } }
  ])
]);
```

**But this fail:**

```ts
test([
  { b: { c: "c" } },
  ...returnAsIt([
    { b: { c: "c" } },
    { b: { b: { c: "c" } } }
  ])
]);
```

```shell
Type '{ b: { c: string; b?: undefined; }; } | { b: { b: { c: string; }; c?: undefined; }; }' is not assignable to type 'Config'.
  Type '{ b: { c: string; b?: undefined; }; }' is not assignable to type 'Config'.
    Property 'c' is missing in type '{ b: { c: string; b?: undefined; }; }' but required in type 'ConditionC'
      'c' is declared here.
```

### 🙂 Expected behavior

I expect that all of the cases above pass as they do in TS, since the types are correct.

### Additional information about the issue

VSCode version **1.89.0**
TS version **5.4.5**

This is tested with `checkJs` set as `true`.

**Comment by Andarist** [comment]
It behaves the same way but you need to make sure to enable `strictNullChecks` for your JS file: [TS Playground](https://www.typescriptlang.org/play/?strictNullChecks=true&ts=5.4.5&filetype=js#code/PQKhCgAIUgBAXAngBwKYBNUDNIG8D2ARgFaoDG8AvpAML4B26AlvEwzVDLMgE75o8keAM7weTegHNqZTnF79UgxHnqoAbkuoBtQgF1OwcOFARocJGkw4CJclVoNmrBgCE53PgKG41mnjpkBuaeisp4dPRYTNKQhIbGph6WGNgRTixs9K6QAD6OjJns1JHRkgkmYB7IAIY8NQC26VEx2noyDGUePKjwAK489MJ4hPj4ADaoNfSUCWQMopDwqIsAvJAAFPMtkgCUkKsAfJBYNePCqADciVUhyw3I4zXLkAAqKKjVdY1472jUdXKIR6-UGw1wf1Qs2gRnAWD69AoWUgIIG9AAgsIAJLwDaA-a4KAo3poyCA66zEzASAAUQAHmgKBhIMI+mQyCthOBlqINtpcHEAFx4SBkYUAIjI4sg1Eoel21x5uP5QpFhGFArFkEl0tlMvl125K2VRIF6pFWp1MplABpTaqzRrRRKpdbZeADYlqfTGct0CcakxxkbeSq5QrIMBqQAFLxKIQAckICcgTGGDTTwgk5SVfMdeFlBsjMbj4STKbTkAzwizUi9tIZ9mZrPZnJDJsgaqdltdsrtnYAdEPUYNMTi+UTO-nNS7dbbJ13FzPtb23UT5R6FUA)

I find it confusing that the TS playground for JS files uses `strict: true, strictNullChecks: false`. When I see `strict: true` set as the default I certainly don't go check all the other options to see if they are not opting out of strictness. This is done here: https://github.com/microsoft/TypeScript-Website/blob/eaa8205658445d4df6c0ca04e42fab1019f71df0/packages/sandbox/src/compilerOptions.ts#L21

**Comment by Arkellys** [comment]
Ah yes you're right, the issue only appears when `strictNullCheck` is `false`. Here is [the updated playground](https://www.typescriptlang.org/play/?ts=5.4.5&filetype=js&strictNullChecks=false#code/PQKhCgAIUgBAXAngBwKYBNUDNIG8D2ARgFaoDG8AvpAML4B26AlvEwzVDLMgE75o8keAM7weTegHNqZTnF79UgxHnqoAbkuoBtQgF1OwcOFARocJGkw4CJclVoNmrBgCE53PgKG41mnjpkBuaeisp4dPRYTNKQhIbGph6WGNgRTixs9K6QAD6OjJns1JHRkgkmYB7IAIY8NQC26VEx2noyDGUePKjwAK489MJ4hPj4ADaoNfSUCWQMopDwqIsAvJAAFPMtkgCUkKsAfJBYNePCqADciVUhyw3I4zXLkAAqKKjVdY1472jUdXKIR6-UGw1wf1Qs2gRnAWD69AoWUgIIG9AAgsIAJLwDaA-a4KAo3poyCA66zEzASAAUQAHmgKBhIMI+mQyCthOBlqINtpcHEAFx4SBkYUAIjI4sg1Eoel21x5uP5QpFhGFArFkEl0tlMvl125K2VRIF6pFWp1MplABpTaqzRrRRKpdbZeADYlqfTGct0CcakxxkbeSq5QrIMBqQAFLxKIQAckICcgTGGDTTwgk5SVfMdeFlBsjMbj4STKbTkAzwizUi9tIZ9mZrPZnJDJsgaqdltdsrtnYAdEPUYNMTi+UTO-nNS7dbbJ13FzPtb23UT5R6FUA). I wish I could use this as a workaround for this bug, but turning on `strictNullCheck` gives me even more errors for things I don't want. 🫤

**Comment by Andarist** [comment]
You can replace your `never`s here with `undefined` and that should fix the problem without you having to turn on `strictNullChecks`: [TS playground](https://www.typescriptlang.org/play/?ts=5.4.5&filetype=js&strictNullChecks=false#code/PQKhCgAIUgBAXAngBwKYBNUDNIG8D2ARgFaoDG8AvpAML4B26AlvEwzVDLMgE75o8keAM7weTegHNqZTnF79UgxHgCujbBIzUA2oQC6nYOHCgI0OEjSYcBEuSq0GzVgwBCc7nwFDc6m1roumSGFl6Kynh09FhM0pCERiZmnlYY2FHOLGz0bpAAPk6M2ezU0bGSSaZgnsgAhjx1ALaZMXE6+jIMFZ48qPCqPPTCeIT4+AA2qHX0lElkDKKQ8KhLALyQABQLbZIAlJBrAHyQWHUTwqgA3Mk1YStNyBN1K5AAKiiotQ3NeB9o1AalTCfQGQxGuH+qDm0GM4Cw6goOUgoMG9AAgsIAJLwTZAg64KAo-poyBAm5zUzASAAUQAHmgKBhIMJVGQyKthOAVqJNjpcAkAFx4SBkYUAIjI4sg1Eo+j2Nx5uP5QpFhGFArFkEl0tlMvlN25q2VRIF6pFWp1MplABpTaqzRrRRKpdbZeADclqfTGSt0Kc6kwJkbeSq5QrIMBqQAFbxKIQAckICcgTBGTTTwgklSVfMdeFlBsjMbjkSTKbTkAzwizUi9tIZDmZrPZnJDJsgaqdltdsrtnYAdEPUUNMTi+UTO-nNS7dbbJ13FzPtb23UT5R6FUA)
